referenceqvm/qvm_stabilizer.py

```diff
@@ -0,0 +1,479 @@
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+##############################################################################
+"""
+Pure QVM that only executes pyQuil programs containing Clifford group generators.
+This is based off of the paper by Aaronson and Gottesman Phys. Rev. A 70, 052328
+"""
+from functools import reduce
+from pyquil.quil import Program, get_classical_addresses_from_program
+from pyquil.quilbase import *
+from pyquil.paulis import PauliTerm, sI, sZ, sX, sY
+from pyquil.wavefunction import Wavefunction
+
+from referenceqvm.qam import QAM
+from referenceqvm.gates import stabilizer_gate_matrix
+from referenceqvm.unitary_generator import value_get, tensor_up
+from referenceqvm.stabilizer_utils import (project_stabilized_state,
+                                           symplectic_inner_product,
+                                           binary_stabilizer_to_pauli_stabilizer)
+
+
+class QVM_Stabilizer(QAM):
+    """
+    A  P Y T H O N
+
+    Q U A N T U M
+
+    V I R T U A L
+
+    M A C H I N E
+
+    S I M U L A T I N G
+
+    S T A B I L I Z E R
+
+    S T A T E S
+    """
+    def __init__(self, num_qubits=None, program=None, program_counter=None,
+                 classical_memory=None, gate_set=stabilizer_gate_matrix,
+                 defgate_set=None):
+        """
+        Subclassed from QAM this is a pure QVM.
+        """
+        super(QVM_Stabilizer, self).__init__(num_qubits=num_qubits, program=program,
+                                             program_counter=program_counter,
+                                             classical_memory=classical_memory,
+                                             gate_set=gate_set,
+                                             defgate_set=defgate_set)
+        # this sets that arbitrary instructions are not allowed
+        # this can probably be factored out eventually.  It is a relic of the
+        # the old days of generating unitaries of programs
+        self.all_inst = False
+        if num_qubits is None:
+            self.tableau = None
+        else:
+            self.tableau = self._n_qubit_tableau(num_qubits)
+
+    def load_program(self, pyquil_program):
+        """
+        Loads a pyQuil program into the QAM memory.
+            Synthesizes pyQuil program into instruction list.
+            Initializes program object and program counter.
+
+        This overwrites the parent class load program.  Required because
+        unitary check QVM checks makes it annoying to allow measurements.
+        TODO: Change load_program for each qvm-subclass
+
+        :param Program pyquil_program: program to be run
+        """
+        # typecheck
+        if not isinstance(pyquil_program, Program):
+            raise TypeError("I can only generate from pyQuil programs")
+
+        if self.all_inst is None:
```

Comment:
this seems out of place

Reply:
There is a better way to do this in terms of ABC's but that will be a future upgrade.  I will add to the issue list.